external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetRangeCalculator.scala

```diff
@@ -61,19 +61,23 @@ private[kafka010] class KafkaOffsetRangeCalculator(val minPartitions: Option[Int
 
       // Splits offset ranges with relatively large amount of data to smaller ones.
       val totalSize = offsetRanges.map(_.size).sum
-      val idealRangeSize = totalSize.toDouble / minPartitions.get
-
       offsetRanges.flatMap { range =>
-        // Split the current range into subranges as close to the ideal range size
-        val numSplitsInRange = math.round(range.size.toDouble / idealRangeSize).toInt
-
-        (0 until numSplitsInRange).map { i =>
-          val splitStart = range.fromOffset + range.size * (i.toDouble / numSplitsInRange)
-          val splitEnd = range.fromOffset + range.size * ((i.toDouble + 1) / numSplitsInRange)
-          KafkaOffsetRange(
-            range.topicPartition, splitStart.toLong, splitEnd.toLong, preferredLoc = None)
+        val tp = range.topicPartition
+        val size = range.size
+        // number of partitions to divvy up this topic partition to
+        val parts = math.max(math.round(size.toDouble / totalSize * minPartitions.get), 1).toInt
```

Comment:
This one ensures we never drop a TopicPartition.

Reply:
The ratio calculation looks good, but `round` seems to generate less partitions. Is there a reason to choose `round` instead of `ceiling`?

Reply:
Yeah I'm seeing the same. Suppose 4 offsetRanges divide 1 partition for each 0.25, then we lost 1. The number of lost partitions may vary.

In other words, if we use ceil, it may overflow the minimum partitions, and the number of exceeding partitions may vary. We don't guarantee for this calculator to return partitions closest to minimum partitions, so it seems OK.

If we really would like to make this strict, we could apply "allocation" - calculating ratio on each offsetRange, and allocate partitions to each offsetRange according to ratio (apply minimum of 1 for safeness), and allocate extra partitions to some offsetRanges if there're remaining partitions. Not sure we would like to deal with complexity.

Reply:
Yep, it's a hint. And when the number of partitions is less than `minPartition`, we will try our best to split. Agreed that the option name `minPartition` is not accurate.

Reply:
Then, could you update the document instead in a more accurate way?

Reply:
@dongjoon-hyun I think the doc for this method is accurate : https://github.com/apache/spark/blob/c4010a25a7311d58c06880c00ad85bfb7f583b90/external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetRangeCalculator.scala#L38

Reply:
I meant [the structured streaming Kafka integration](https://github.com/apache/spark/blame/master/docs/structured-streaming-kafka-integration.md#L391-L400)~

Reply:
A few days ago, `minPartitions` is added to the documentation for master/branch-2.4 via https://github.com/apache/spark/pull/25219 .